sdks/java/extensions/schemaio-expansion-service/src/main/java/org/apache/beam/sdk/extensions/schemaio/expansion/ExternalSchemaIOTransformRegistrar.java

```diff
@@ -0,0 +1,142 @@
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableMap;
+
+@Experimental(Experimental.Kind.PORTABILITY)
+@AutoService(ExternalTransformRegistrar.class)
+public class ExternalSchemaIOTransformRegistrar implements ExternalTransformRegistrar {
+
+  @Override
+  public Map<String, ExternalTransformBuilder<?, ?, ?>> knownBuilderInstances() {
+    ImmutableMap.Builder builder = ImmutableMap.<String, ExternalTransformRegistrar>builder();
+    try {
+      for (SchemaIOProvider schemaIOProvider : ServiceLoader.load(SchemaIOProvider.class)) {
+        builder.put(
+            "beam:external:java:" + schemaIOProvider.identifier() + ":read:v1",
+            new ReaderBuilder(schemaIOProvider));
+        builder.put(
+            "beam:external:java:" + schemaIOProvider.identifier() + ":write:v1",
+            new WriterBuilder(schemaIOProvider));
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e.getMessage());
+    }
+    return builder.build();
+  }
+
+  public static class Configuration {
+    String location = "";
+    byte[] config = new byte[0];
+    @Nullable byte[] dataSchema = null;
+
+    public void setLocation(String location) {
+      this.location = location;
+    }
+
+    public void setConfig(byte[] config) {
+      this.config = config;
+    }
+
+    public void setDataSchema(byte[] dataSchema) {
+      this.dataSchema = dataSchema;
+    }
+  }
+
+  @Nullable
+  private static Schema translateSchema(@Nullable byte[] schemaBytes) throws Exception {
+    if (schemaBytes == null) {
+      return null;
+    }
+    SchemaApi.Schema protoSchema = SchemaApi.Schema.parseFrom(schemaBytes);
+    return SchemaTranslation.schemaFromProto(protoSchema);
+  }
+
+  private static Row translateRow(byte[] rowBytes, Schema configSchema) throws Exception {
+    RowCoder rowCoder = RowCoder.of(configSchema);
+    InputStream stream = new ByteArrayInputStream(rowBytes);
+    return rowCoder.decode(stream);
+  }
+
+  private static class ReaderBuilder
+      implements ExternalTransformBuilder<Configuration, PBegin, PCollection<Row>> {
+    SchemaIOProvider schemaIOProvider;
+
+    ReaderBuilder(SchemaIOProvider schemaIOProvider) {
+      this.schemaIOProvider = schemaIOProvider;
+    }
+
+    @Override
+    public PTransform<PBegin, PCollection<Row>> buildExternal(Configuration configuration) {
+      try {
+        return schemaIOProvider
+            .from(
+                configuration.location,
+                translateRow(configuration.config, schemaIOProvider.configurationSchema()),
+                translateSchema(configuration.dataSchema))
+            .buildReader();
+      } catch (Exception e) {
+        throw new RuntimeException("Could not convert configuration proto to row or schema.");
+      }
+    }
+  }
+
+  private static class WriterBuilder
```

Comment:
There should probably be some unit tests that exercise WriterBuilder and ReaderBuilder. You could create instances that reference a mock SchemaIOProvider and verify they do what we expect.

Reply:
ACK, I will add these in.